modules/ardrivo/test/Wire.cxx

```diff
@@ -0,0 +1,176 @@
+}
+
+void transfer_test(std::uint8_t bus_no, std:: uint8_t slave_no){
+    const auto chunk_size = GENERATE(take(256 / sizeof(uint8_t), random(1, 31)));
+    const std::vector slave_in = GENERATE_COPY(
+        take(1, chunk(chunk_size, random<uint8_t>(std::numeric_limits<std::uint8_t>::min(), std::numeric_limits<std::uint8_t>::max()))));
+    std::vector<decltype(slave_in)::value_type> slave_out(slave_in.size());
+    const std::vector master_in = GENERATE_COPY(
+        take(1, chunk(chunk_size, random<uint8_t>(std::numeric_limits<std::uint8_t>::min(), std::numeric_limits<std::uint8_t>::max()))));
+    std::vector<decltype(slave_in)::value_type> master_out{};
+
+    TwoWire test;
+    auto& blocking_bus = std::get<BlockingBus<std::tuple<>>>(board_data->i2c_buses[bus_no].slaves[slave_no].second);
+    auto& packet_buffer = board_data->i2c_buses[bus_no].slaves[slave_no].first;
+    auto& bus_mut = board_data->i2c_buses[bus_no].devices_mut;
+    board_data->i2c_buses[bus_no].slaves[slave_no].second.emplace<0>();
+    auto slave_thread = std::thread(slave_worker, slave_in.data(), slave_out.data(), (size_t)slave_in.size(), std::ref(packet_buffer), std::ref(blocking_bus), std::ref(bus_mut));
+
+    test.begin(bus_no, bus_no);
+    test.beginTransmission(slave_no);
+    test.write(reinterpret_cast<const std::uint8_t*>(master_in.data()), master_in.size());
+    test.endTransmission();
+
+    test.requestFrom(slave_no, (size_t)slave_in.size(), false);
+    slave_thread.join();
+    while (test.available()) {
+        master_out.push_back(test.read());
+    }
+    REQUIRE(ranges::equal(master_in, slave_out));
+    REQUIRE(ranges::equal(slave_in, master_out));
+}
+
+TEST_CASE("Initiate Wire with or without address", "[Wire Begin]") {
+    TwoWire test;
+    constexpr std::uint8_t address = 64;
+    init_fake();
+
+    // Valid cases
+    REQUIRE_NOTHROW(test.begin());
+    REQUIRE_NOTHROW(test.begin(address));
+}
+
+TEST_CASE("Initiate Wire with sda and scl", "[Wire begin sda&scl]") {
+    TwoWire test;
+    constexpr std::uint8_t address = 64;
+    init_fake();
+
+    // Valid case
+    REQUIRE_NOTHROW(test.begin(1, 1));
+    REQUIRE_NOTHROW(test.begin(1, 1, address));
+    // SDA or SCL doesn't exist
+    REQUIRE_THROWS_AS(test.begin(101, 1), std::runtime_error);
+    REQUIRE_THROWS_AS(test.begin(1, 101), std::runtime_error);
+    // SDA SCL pair are not an I2C channel
+    REQUIRE_THROWS_AS(test.begin(3, 3), std::runtime_error);
+    REQUIRE_THROWS_AS(test.begin(3, 3, address), std::runtime_error);
+    // Invalid address
+    REQUIRE_THROWS_AS(test.begin(1, 1, 128), std::runtime_error);
+}
+
+TEST_CASE("Request the data from slave Generator", "[Wire RequestFrom generator]") {
+    const auto chunk_size = GENERATE(take(256 / sizeof(uint8_t), random(1, 31)));
+    const std::vector slave_in = GENERATE_COPY(
+        take(1, chunk(chunk_size, random<uint8_t>(std::numeric_limits<std::uint8_t>::min(), std::numeric_limits<std::uint8_t>::max()))));
```

Comment:
`random<uint8_t>` has to be `random<unsigned>` for windows to build iirc

Reply:
Then it doesn't work in the others as we tested yesterday.

Reply:
Windows doesnt allow uint8_t to generate random
```\include\random(1902,1): error C2338: note: char, signed char, unsigned char, char8_t, int8_t, and uint8_t are not allowed ```
Will try to make a patch for it

Reply:
The says that and MSVC just happens to be the only compliant compiler on this absolutely awful detail

Reply:
Fixed it with dcb16b9, hopefully the fix isn't stupid :)